app/jobs/results_recorder_job.rb

```diff
@@ -1,41 +1,51 @@
+# Preconditions:
+# PlexerJob has made a matching ArchivePreservedCopyPart row
+#
 # Responsibilities:
 # Update DB per event info.
 # Is this event the last needed for the DV to be complete?
 # If NO, do nothing further.
 # If YES, send a message to a non-job pub/sub queue.
 class ResultsRecorderJob < ApplicationJob
   queue_as :endpoint_events
-  attr_accessor :pc, :pcs
+  attr_accessor :apc, :apcs
 
   before_perform do |job|
-    job.pcs ||= PreservedCopy
-                .by_druid(job.arguments.first)
-                .joins(endpoint: [:endpoint_type])
-                .where(
-                  endpoint_types: { endpoint_class: 'archive' },
-                  preserved_copies: { version: job.arguments.second }
-                )
-    job.pc ||= pcs.find_by!('endpoints.delivery_class' => Object.const_get(job.arguments.third))
+    job.apcs ||= ArchivePreservedCopy
+                 .by_druid(job.arguments.first)
+                 .joins(:archive_endpoint)
+                 .where(version: job.arguments.second)
+    job.apc ||= apcs.find_by!('archive_endpoints.delivery_class' => Object.const_get(job.arguments.fourth))
   end
 
   # @param [String] druid
   # @param [Integer] version
+  # @param [String] s3_part_key
   # @param [String] delivery_class Name of the worker class that performed delivery
-  def perform(druid, version, _delivery_class)
-    raise "Status shifted underneath replication: #{pc.inspect}" unless pc.unreplicated?
-    pc.ok!
-    return unless pcs.reload.all?(&:ok?)
+  def perform(druid, version, s3_part_key, _delivery_class)
+    part = apc_part!(s3_part_key)
+    part.ok!
+    apc.ok! if part.all_parts_replicated?
```

Comment:
what about a transaction here?  i definitely wouldn't hold up the PR over it.

would it be possible/plausible for a part to leave the replicated state?  i guess i can imagine that happening because of an audit check.

mostly it seemed to me like if `apc.ok!` is being set based on the result of the query (queries?) in `all_parts_replicated?`, then that stuff should happen within a single consistent DB state.

Reply:
I don't see them as interdependent, only order-dependent.  `part.ok!` is still true even if `apc.ok!` fails, and we are still "ahead" to have recorded that much of the result.